### PR TITLE
fix(udf): fix "udf returned no data" error

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR fixes a bug that caused "UDF service returned no data" error when calling external UDF on an empty chunk (all rows are invisible). Now it will return an empty array without sending a request to UDF server.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xzhseh** [comment]
QQ: could we have some test cases for this scenario?

**Comment by wangrunji0408** [comment]
> QQ: could we have some test cases for this scenario?

Actually I failed to reproduce this bug in an end-to-end manner, which requires an input chunk with 0 cardinality in the projection executor. The following query doesn't work because the filter executor won't output chunk if all rows are filtered out. 

```sql
create table t(a int);
insert into t values (1), (2);
select udf() from t where a = 0;
```

**Comment by xzhseh** [comment]
> > QQ: could we have some test cases for this scenario?
> 
> 
> 
> Actually I failed to reproduce this bug in an end-to-end manner, which requires an input chunk with 0 cardinality in the projection executor. The following query doesn't work because the filter executor won't output chunk if all rows are filtered out. 
> 
> 
> 
> ```sql
> 
> create table t(a int);
> 
> insert into t values (1), (2);
> 
> select udf() from t where a = 0;
> 
> ```

I'm curious how initially this bug is detected. Otherwise it seems that it's not possible to occur in normal context if output a empty chunk can't be simulated. 🤔

**Comment by KeXiangWang** [comment]
> how initially this bug is detected

Discovered in production environment. We found metrics that said many UDF calls `udf returned no data`.

File: src/expr/core/src/expr/expr_udf.rs [hunk 1]
```diff
@@ -89,51 +96,29 @@ impl Expression for UserDefinedFunction {
...
-            .collect();
-        let opts = arrow_array::RecordBatchOptions::default()
-            .with_row_count(Some(compacted_chunk.capacity()));
-        let input = arrow_array::RecordBatch::try_new_with_options(
-            self.arg_schema.clone(),
-            compacted_columns,
-            &opts,
-        )
-        .expect("failed to build record batch");
+    async fn eval_inner(&self, input: &DataChunk) -> Result<ArrayRef> {
```
Comment (xxchan) [inline src/expr/core/src/expr/expr_udf.rs:105]: add a comment that it returns error if result chunk is empty (incl. empty input)?

File: src/expr/udf/src/external.rs [hunk 2]
```diff
@@ -139,21 +139,17 @@ impl ArrowFlightUdfClient {
...
-        let remaining = output_stream.try_collect::<Vec<_>>().await?;
-        if remaining.is_empty() {
-            Ok(head)
-        } else {
-            Ok(arrow_select::concat::concat_batches(
-                &head.schema(),
-                std::iter::once(&head).chain(remaining.iter()),
-            )?)
+        let mut output_stream = self
+            .call_stream_internal(id, stream::once(async { input }))
```
Comment (xxchan) [inline src/expr/udf/src/external.rs:143]: Why to change this? `call_stream` -> `call_stream_internal`

File: src/expr/core/src/expr/expr_udf.rs [hunk 3]
```diff
@@ -72,13 +73,19 @@ impl Expression for UserDefinedFunction {
...
+            return Ok(builder.finish().into_ref());
+        }
         let mut columns = Vec::with_capacity(self.children.len());
         for child in &self.children {
             let array = child.eval(input).await?;
             columns.push(array);
         }
-        self.eval_inner(columns, vis).await
+        let chunk = DataChunk::new(columns, input.visibility().clone());
+        self.eval_inner(&chunk).await
```
Comment (xxchan) [inline src/expr/core/src/expr/expr_udf.rs:88]:
```suggestion
        self.eval_inner(&input).await
```
?

File: src/expr/core/src/expr/expr_udf.rs [hunk 4]
```diff
@@ -44,6 +44,7 @@ pub struct UserDefinedFunction {
     children: Vec<BoxedExpression>,
     arg_types: Vec<DataType>,
     return_type: DataType,
+    #[allow(dead_code)]
```
Comment (xxchan) [inline src/expr/core/src/expr/expr_udf.rs:47]:
```suggestion
    #[expect(dead_code)]
```

File: src/expr/udf/src/external.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/udf/src/external.rs:143]: If the input is empty, `call_stream` (under the hood is `do_exchange`) will return no `RecordBatch`. But we expect an empty `RecordBatch` in `call_internal`. The difference is that empty chunks contains schema info. So we introduced `call_stream_internal`, which returns an `FlightRecordBatchStream` that can be converted to an empty chunk.

File: src/expr/core/src/expr/expr_udf.rs (hunk 1 above)
Comment (xxchan) [reply src/expr/core/src/expr/expr_udf.rs:105]: Hmmm, isn't the modification of `eval` for this?

File: src/expr/udf/src/external.rs (hunk 2 above)
Comment (xxchan) [reply src/expr/udf/src/external.rs:143]: I don't understand. Isn't `call_stream` directly calls `call_stream_internal`..?

File: src/expr/core/src/expr/expr_udf.rs (hunk 1 above)
Comment (xxchan) [reply src/expr/core/src/expr/expr_udf.rs:105]: Do you mean the changes of external.rs can fix it, but the `eval` change is just another optimization?

Comment (wangrunji0408) [reply src/expr/core/src/expr/expr_udf.rs:105]: Exactly!

File: src/expr/udf/src/external.rs (hunk 2 above)
Comment (xxchan) [reply src/expr/udf/src/external.rs:143]: Oh, I understood. `call_stream` make the type opaque

Comment (xxchan) [reply src/expr/udf/src/external.rs:143]: I think just letting `call_stream` returns `FlightRecordBatchStream` is enough. No need to have a separate function returning `impl Stream`

**Review by xxchan — APPROVED** [review]

File: src/expr/udf/src/external.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/udf/src/external.rs:143]: Yes. Here is the call graph:

```mermaid
graph TB
    ScalarFunctions --> Call
    Call --> Call_internal
    Call_internal --> Call_stream_internal
    TableFunctions --> Call_stream
    Call_stream --> Call_stream_internal

```

Comment (wangrunji0408) [reply src/expr/udf/src/external.rs:143]: > I think just letting `call_stream` returns `FlightRecordBatchStream` is enough. No need to have a separate function returning `impl Stream`

I just don't want to expose arrow-specific type `FlightRecordBatchStream` to the caller. 😄

**Pull request merged**